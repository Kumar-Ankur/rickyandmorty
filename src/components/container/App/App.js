import React, { Component } from "react";
import { connect } from "react-redux";
import * as actionCreator from "../../../actions";
import * as constants from "../../../constants";
import Accordion from "../../presentation/Accordion/Accordion";
import SortButton from "../SortButton/SortButton";
import SpeciesView from "../SpeciesView/SpeciesView";
import GenderView from "../GenderView/GenderView";
import OriginView from "../OriginView/OriginView";
import SearchView from "../SearchView/SearchView";
import ListItems from "../../presentation/ListItems/ListItems";
import Footer from "../../presentation/Footer/Footer";
import "./App.scss";

class App extends Component {
  componentDidMount() {
    // dispatch action to get data from API
    this.props.getCharacters();
  }

/**
 * @function getSearchData
 * @param { Object } filteredOriginItem
 * @description - function to filtered out data based on search text
 * @returns { Object } - return the filtered data based on search text by user
 */
  getSearchData = (filteredOriginItem) => {
    const { search_text } = this.props;
    let filteredSearchItem =
      search_text.length > 0
        ? filteredOriginItem.filter((item) => {
            return item.name.toLowerCase().includes(search_text.toLowerCase());
          })
        : filteredOriginItem;

    return filteredSearchItem;
  };

/**
 * @function getOriginFilter
 * @param { Object } filteredGenderItem
 * @description - function to filtered out data based on origin 
 * @returns { Object } - return the filtered data based on Origin value
 */
  getOriginFilter = (filteredGenderItem) => {
    const { filter_origin_data } = this.props;
    const filteredOriginItem =
      filter_origin_data.length > 0
        ? filteredGenderItem.filter((item) => {
            return filter_origin_data.indexOf(item.origin.name) >= 0;
          })
        : filteredGenderItem;

    return this.getSearchData(filteredOriginItem);
  };

/**
 * @function getGenderFilter
 * @param { Object } filteredItem
 * @description - function to filtered out data based on gender 
 * @returns { Object } - return the filtered data based on gender value
 */
  getGenderFilter = (filteredItem) => {
    const { filter_gender_data } = this.props;
    let filteredGenderItem =
      filter_gender_data.length > 0
        ? filteredItem.filter((item) => {
            return filter_gender_data.indexOf(item.gender) >= 0;
          })
        : filteredItem;

    return this.getOriginFilter(filteredGenderItem);
  };

/**
 * @function getSpeciesFilter
 * @param { Object } all_characters
 * @param { Array } filter_data
 * @description - function to filtered out data based on species 
 * @returns { Object } - return the filtered data based on species value
 */
  getSpeciesFilter = (all_characters, filter_data) => {
    let filteredItem =
      filter_data.length > 0
        ? all_characters.filter((item) => {
            return filter_data.indexOf(item.species) >= 0;
          })
        : all_characters;

    return this.getGenderFilter(filteredItem);
  };

/**
 * @function getFilteredItem
 * @description - function to return final filtered data based on species, origin, gender and search text.
 */
  getFilteredItem = () => {
    const { all_characters, filter_data } = this.props;
    return this.getSpeciesFilter(all_characters, filter_data);
  };

  render() {
    let filteredItem = this.getFilteredItem();

    return (
      <div className="main">
        <div className="main_heading">{constants.APP_HEADING}</div>
        <div className="sortBox">
          <SortButton />
        </div>
        <div className="main_functionality">
          <div className="main_functionality-species">
            <Accordion
              title={ constants.ACCORDION_TEXT}
              content={[<SpeciesView />, <GenderView />, <OriginView />]}
            />
          </div>
        </div>

        <div className="main_input">
          <SearchView />
        </div>

        <div className="viewcontent">
          <ListItems filteredItem={filteredItem} />
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    all_characters: state.all_characters,
    filter_data: state.filter_data,
    filter_gender_data: state.filter_gender_data,
    filter_origin_data: state.filter_origin_data,
    search_text: state.search_text,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    getCharacters: () => dispatch(actionCreator.getCharacters()),
    sortAscendingById: () => dispatch(actionCreator.sortAscendingById()),
    sortDescendingById: () => dispatch(actionCreator.sortDescendingById()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
