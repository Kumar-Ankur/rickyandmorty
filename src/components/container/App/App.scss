.main {
  background: linear-gradient(to right, #3e5151, #decba4);
  width: 100vw;
  position: relative;
  top: 0;
  color: white;

  &_heading {
    text-align: center;
    position: relative;
    top: 2rem;
    font-size: 2rem;
    text-transform: uppercase;
    font-weight: 600;
    letter-spacing: 2px;
    font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
  }

  &_functionality {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    transition: ease-in-out 1s all;
    padding-top: 5rem;

    &-species {
      width: 100%;
    }
  }

  &_input {
    display: flex;
    justify-content: center;
    margin-top: 2rem;
  }
}

.viewcontent {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  transition: ease-in-out 1s all;
  position: relative;
  top: 9rem;
}

.sortBox {
  position: absolute;
  right: 2rem;
  top: 2rem;
}
