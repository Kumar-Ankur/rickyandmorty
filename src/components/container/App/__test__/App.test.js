import React from "react";
import { shallow } from "enzyme";
import { applyMiddleware, createStore } from "redux";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "../../../../reducers";
import App from "../App";

const middleware = createSagaMiddleware();
const mockStore = createStore(rootReducer, applyMiddleware(middleware));
/**
 * Factory function to create a shallowwrapper for the App component.
 * @function setUp
 * @param { Object } props - Component props specific to this setup
 * @param { Object } state - Initial state for setUp
 * @returns { shallowWrapper }
 *
 */

function setUp(props = {}) {
  const wrapper = shallow(<App store={mockStore} {...props} />);
  return wrapper;
}

/**
 * Return shallowWarpper containing node(s) with the given data-test attribute
 * @function findByTestAttr
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper to search within
 * @param {string} val - value of data-test attribute for search
 * @returns { shallowWrapper}
 */
function findByTestAttr(wrapper, val) {
  return wrapper.find(`[data-test="${val}"]`);
}

describe("App component", () => {
  test("snapshot testing for app component", () => {
    const wrapper = setUp();
    expect(wrapper).toMatchSnapshot();
  });

  test("render app without crash", () => {
    const wrapper = setUp().dive().dive();
    const appComponent = findByTestAttr(wrapper, "component-app");
    expect(appComponent.length).toBe(1);
  });

  test("sort component render without crash", () => {
    const wrapper = setUp().dive().dive();
    const sortComponent = findByTestAttr(wrapper, "component-sort");
    expect(sortComponent.length).toBe(1);
  });

  test("accodion component render without crash", () => {
    const wrapper = setUp().dive().dive();
    const accordionComponent = findByTestAttr(wrapper, "component-accordion");
    expect(accordionComponent.length).toBe(1);
  });

  test("search component render without crash", () => {
    const wrapper = setUp().dive().dive();
    const searchComponent = findByTestAttr(wrapper, "component-search");
    expect(searchComponent.length).toBe(1);
  });

  test("list component render without crash", () => {
    const wrapper = setUp().dive().dive();
    const listComponent = findByTestAttr(wrapper, "component-content");
    expect(listComponent.length).toBe(1);
  });
});

describe("function condition validate inside app component", () => {
  let wrapper;
  let instance;
  const getCharactersMock = jest.fn();
  const props = {
    getCharacters: getCharactersMock
  }
  beforeEach(() => {
    wrapper = setUp(props).dive().dive();
    instance = wrapper.instance();
  });

  test('check `getCharacters` call on componentDidMount', () => {
    const getCharactersMock = jest.fn();
    instance.getCharacters = getCharactersMock;
    instance.getCharacters();
    expect(getCharactersMock).toHaveBeenCalled(); 
  });

  test("check `getFilteredItem` call without error", () => {
    const getSpeciesFilterMock = jest.fn();
    instance.getSpeciesFilter = getSpeciesFilterMock;
    instance.getFilteredItem();
    expect(getSpeciesFilterMock).toHaveBeenCalled();
  });

  test("check `getSpeciesFilter` call without error", () => {
    const getGenderFilterMock = jest.fn();
    instance.getGenderFilter = getGenderFilterMock;
    const all_characters = [];
    const filter_data = [];
    instance.getSpeciesFilter(all_characters, filter_data);
    expect(getGenderFilterMock).toHaveBeenCalled();
  });

  test('check `getGenderFilter` call without error', () => {
    const getOriginFilterMock = jest.fn();
    instance.getOriginFilter = getOriginFilterMock;
    const filteredItem = [];
    instance.getGenderFilter(filteredItem);
    expect(getOriginFilterMock).toHaveBeenCalled();
  });

  test('check `getOriginFilter` call without error', () => {
      const getSearchDataMock = jest.fn();
      instance.getSearchData = getSearchDataMock;
      instance.getOriginFilter();
      expect(getSearchDataMock).toHaveBeenCalled();
  });

  test('check `getSearchData` call without error', () => {
      const getSearchDataMock = jest.fn();
      instance.getSearchData = getSearchDataMock;
      instance.getSearchData();
      expect(getSearchDataMock).toHaveBeenCalled();
  })
});
