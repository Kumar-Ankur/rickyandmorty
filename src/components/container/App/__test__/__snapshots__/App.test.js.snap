// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot testing for app component 1`] = `ShallowWrapper {}`;
